### PR TITLE
[MV-378] Impl GLU opeartion (forward only, contiguous)

### PR DESCRIPTION
**Pull request by cognaiger9**
Checklist

- [x] Port kernel from OpenCL to HIP
- [x] CPU implementation of splitting input tensor and performing operation
- [x] Implement driver (float, fp16, bfp16)
- [x] Implement gtest and pass
- [x] Benchmark performance compared to MoDNN

<details close>
<summary>float16</summary>
<br>

| op_name | dtype   | size           | dim | direction | ROCm pytorch | MIOpen HIP | Improvement |
|---------|---------|----------------|-----|-----------|--------------|------------|-------------|
| GLU     | float16 | [8 120 1]      | 0   | fwd       | 6752         | 4071       | 1.66        |
| GLU     | float16 | [8 1023 1]     | 0   | fwd       | 7632         | 4888       | 1.56        |
| GLU     | float16 | [8 1024 768]   | 0   | fwd       | 28078        | 36194      | 0.78        |
| GLU     | float16 | [16 1024 768]  | 0   | fwd       | 51755        | 68229      | 0.76        |
| GLU     | float16 | [48 8 512 512] | 0   | fwd       | 386416       | 537105     | 0.72        |

</details>

<details close>
<summary>float32</summary>
<br>

| op_name | dtype   | size           | dim | direction | ROCm pytorch | MIOpen HIP | Improvement |
|---------|---------|----------------|-----|-----------|--------------|------------|-------------|
| GLU     | float32 | [8 120 1]      | 0   | fwd       | 5743         | 3911       | 1.47        |
| GLU     | float32 | [8 1023 1]     | 0   | fwd       | 6224         | 5048       | 1.23        |
| GLU     | float32 | [8 1024 768]   | 0   | fwd       | 32669        | 35127      | 0.93        |
| GLU     | float32 | [16 1024 768]  | 0   | fwd       | 61835        | 66362      | 0.93        |
| GLU     | float32 | [48 8 512 512] | 0   | fwd       | 455882       | 509159     | 0.9        |

</details>

<details close>
<summary>bfloat16</summary>
<br>

| op_name | dtype   | size           | dim | direction | ROCm pytorch | MIOpen HIP | Improvement |
|---------|---------|----------------|-----|-----------|--------------|------------|-------------|
| GLU     | bfloat16 | [8 120 1]      | 0   | fwd       | 6496         | 3928       | 1.65        |
| GLU     | bfloat16 | [8 1023 1]     | 0   | fwd       | 7967         | 4871       | 1.64        |
| GLU     | bfloat16 | [8 1024 768]   | 0   | fwd       | 29694        | 36105      | 0.82        |
| GLU     | bfloat16 | [16 1024 768]  | 0   | fwd       | 54795        | 69082      | 0.79        |
| GLU     | bfloat16 | [48 8 512 512] | 0   | fwd       | 412382       | 530351     | 0.78        |

</details>

* Average over all cases:

|type    |average|
|--------|-------|
|float16 | 1.09  |
|float32 | 1.09  |
|bfloat16| 1.14  |

### CODE REVIEW

File: include/miopen/miopen.h [hunk 1]
```diff
@@ -6582,6 +6582,39 @@ MIOPEN_EXPORT miopenStatus_t miopenBackendInitialize(miopenBackendDescriptor_t d
 // CLOSEOUT BackendAPI DOXYGEN GROUP
 #endif // MIOPEN_BETA_API
 
+#ifdef MIOPEN_BETA_API
+
+// GLU APIs
+/** @addtogroup glu
```
Comment (seungmanhan) [inline include/miopen/miopen.h:6589]: Add * @defgroup glu at the beginning of miopen.h.

File: src/glu.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
```
Comment (seungmanhan) [inline src/glu.cpp]: 2023 -> 2024 Please check other codes as well.

File: src/glu.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+#include <miopen/glu/solvers.hpp>
+#include <miopen/glu.hpp>
+#include <miopen/tensor.hpp>
+
+namespace miopen {
+
+miopenStatus_t GLUForward(Handle& handle,
+                          const TensorDescriptor& inputDesc,
+                          const TensorDescriptor& inputSplitDesc,
+                          Data_t a,
```
Comment (seungmanhan) [inline src/glu.cpp]: Please use the tensor name and description uniformly, such as a,b or x,y or input, output, etc.

File: src/kernels/MIOpenGLU.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+
+__device__ FLOAT_ACCUM sigmoid(FLOAT_ACCUM x) { return 1.0f / (1.0f + exp(-x)); }
+
+extern "C" __global__ void GLUFwdContiguous(const FLOAT* __restrict__ a,
```
Comment (seungmanhan) [inline src/kernels/MIOpenGLU.cpp]: Please refer to https://github.com/moreh-dev/MIOpen/pull/2#discussion_r1562014234 and modify it.

File: test/gtest/glu.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+{
+    const auto& tmp = miopen::GetStringEnv(ENV(MIOPEN_TEST_FLOAT_ARG));
+    if(tmp.empty())
+    {
+        return "";
+    }
+    return tmp;
+}
+
+struct GLUTestFloat : GLUTest<float>
```
Comment (seungmanhan) [inline test/gtest/glu.cpp:46]: add tests for half and bfloat16.

File: test/gtest/glu.hpp [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                                    output_dev.get());
+
+        EXPECT_EQ(status, miopenStatusSuccess);
+
+        output.data = handle.Read<T>(output_dev, output.data.size());
+    }
+
+    void Verify()
+    {
+        double threshold = std::numeric_limits<T>::epsilon();
```
Comment (seungmanhan) [inline test/gtest/glu.hpp]: It is better to compare errors the same as driver.

File: include/miopen/miopen.h (hunk 1 above)
Comment (cognaiger9) [reply include/miopen/miopen.h:6589]: I have added * @defgroup glu miopen.h at [66bd845](https://github.com/moreh-dev/MIOpen/pull/6/commits/66bd84579f508fe2d2deb6124c3c962f9bbc2a83)

File: src/glu.cpp (hunk 2 above)
Comment (cognaiger9) [reply src/glu.cpp]: I have changed year in all newly added files to 2024 at [66bd845](https://github.com/moreh-dev/MIOpen/pull/6/commits/66bd84579f508fe2d2deb6124c3c962f9bbc2a83)

File: src/glu.cpp (hunk 3 above)
Comment (cognaiger9) [reply src/glu.cpp]: I have changed tensor name to input/output convention at [f03c628](https://github.com/moreh-dev/MIOpen/pull/6/commits/f03c62870a10ca371a6ce5d161fb59a306f3904b)

File: src/kernels/MIOpenGLU.cpp (hunk 4 above)
Comment (cognaiger9) [reply src/kernels/MIOpenGLU.cpp]: I have changed tensor name to input/output convention at [f03c628](https://github.com/moreh-dev/MIOpen/pull/6/commits/f03c62870a10ca371a6ce5d161fb59a306f3904b)

File: test/gtest/glu.cpp (hunk 5 above)
Comment (cognaiger9) [reply test/gtest/glu.cpp:46]: I have added test for half and bfloat16 at [b5ff947](https://github.com/moreh-dev/MIOpen/pull/6/commits/b5ff94713f992ee78b258e7fb5999f49702c5c41)

File: test/gtest/glu.hpp (hunk 6 above)
Comment (cognaiger9) [reply test/gtest/glu.hpp]: I have defined threshold based on data type (same as driver) at [db96a77](https://github.com/moreh-dev/MIOpen/pull/6/commits/db96a7773cf3de17a7ba75fa91a2b2859497c4e1)

File: driver/main.cpp [hunk 7]
```diff
@@ -72,6 +72,7 @@ int main(int argc, char* argv[])
         std::cout << "ParseCmdLineArgs() FAILED, rc = " << rc << std::endl;
         return rc;
     }
+
```
Comment (kyeonghwanryu) [inline driver/main.cpp]: Please avoid to add irrelevant modification to the pull request, especially when we don't have the ownership of this project.

File: include/miopen/miopen.h [hunk 8]
```diff
@@ -6582,6 +6583,39 @@ MIOPEN_EXPORT miopenStatus_t miopenBackendInitialize(miopenBackendDescriptor_t d
...
+
+/*! @brief Execute a GLU forward contiguous layer
+ *
+ * @param handle                   MIOpen handle (input)
+ * @param inputDesc                Tensor descriptor for data input tensor x (input)
+ * @param inputSplitDesc           Tensor descriptor for splitted data input tensor x (input)
+ * @param inputFirstHalf           First half input data tensor  (input)
+ * @param inputSecondHalf          Second half input data tensor (input)
+ * @param dim                      Dimensions to split. (input)
+ * @param outputDesc               Tensor descriptor for output data tensor y (input)
```
Comment (kyeonghwanryu) [inline include/miopen/miopen.h]: Shouldn't this be output?

File: src/kernels/MIOpenGLU.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+extern "C" __global__ void GLUFwdContiguous(const FLOAT* __restrict__ inputFirstHalf,
+                                            const FLOAT* __restrict__ inputSecondHalf,
+                                            FLOAT* __restrict__ output,
+                                            long N)
+{
+    const size_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+    if(gid >= N)
+        return;
+
+    FLOAT_ACCUM val1 = CVT_FLOAT2ACCUM(inputFirstHalf[gid]);
```
Comment (kyeonghwanryu) [inline src/kernels/MIOpenGLU.cpp:45]: Is there a reason you used FLOAT_ACCUM?
When we accumulate low-precision floating point values, we convert the values to FLOAT_ACCUM to lower the error. (And convert back to FLOAT before it goes to output)
But i don't see that case fits in this situation.

Comment (cognaiger9) [reply src/kernels/MIOpenGLU.cpp:45]: After experimentation, I noticed that omitting the use of FLOAT_ACCUM leads to the failure of the bfp16 test in the driver, whereas only float and fp16 tests pass. Therefore, I've used FLOAT_ACCUM here to ensure that all three data types are covered.

File: driver/main.cpp (hunk 7 above)
Comment (cognaiger9) [reply driver/main.cpp]: I have removed modification at [6f6fe652](https://github.com/moreh-dev/MIOpen/pull/6/commits/6f6fe652a6f21e1e315ea829f9d3927e5a4085e2)

File: include/miopen/miopen.h (hunk 8 above)
Comment (cognaiger9) [reply include/miopen/miopen.h]: I have changed comments at [6f6fe652](https://github.com/moreh-dev/MIOpen/pull/6/commits/6f6fe652a6f21e1e315ea829f9d3927e5a4085e2)

File: src/kernels/MIOpenGLU.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+
+__device__ FLOAT_ACCUM sigmoid(FLOAT_ACCUM x) { return 1.0f / (1.0f + exp(-x)); }
+
+template <typename TI, typename TO>
+__device__ void GLUFwdCOntiguousKernel(const TI* __restrict__ inputFirstHalf,
```
Comment (seungmanhan) [inline src/kernels/MIOpenGLU.cpp:36]: The kernel called GLU is basically an operator that receives input and output tensors one by one, receives dimention as a parameter.
And divides it in half based on dimention , then multiplies the first half of the divided input by the second half to which sigmoid is applied.
Since the work we do is creating replicas of pytorch or tensorflow, be sure to refer to the definitions of pytorch and tensorflow. (https://pytorch.org/docs/stable/generated/torch.nn.GLU.html) 
The API shared externally in miopen.h is used directly by users. Do not change the definition except in some unavoidable cases.
Therefore, in order to implement GLU in the same way as before, you must understand tensor view and slice. See https://github.com/ROCm/MIOpen/pull/2883. Please note that this PR is currently under review, so some details may change.

File: include/miopen/miopen.h [hunk 11]
```diff
@@ -6582,6 +6583,39 @@ MIOPEN_EXPORT miopenStatus_t miopenBackendInitialize(miopenBackendDescriptor_t d
...
+ * @param inputDesc                Tensor descriptor for data input tensor (input)
+ * @param inputSplitDesc           Tensor descriptor for splitted data input tensor (input)
+ * @param inputFirstHalf           First half input data tensor  (input)
+ * @param inputSecondHalf          Second half input data tensor (input)
+ * @param dim                      Dimension to split (input)
+ * @param outputDesc               Tensor descriptor for output data tensor (input)
+ * @param output                   Output data tensor (output)
+ * @return                         miopenStatus_t
+ */
+MIOPEN_EXPORT miopenStatus_t miopenGLUForward(miopenHandle_t handle,
```
Comment (seungmanhan) [inline include/miopen/miopen.h:6604]: GLU is basically an operator that receives only input and output tensors not inputFirstHalf and inputSecondhalf and output, receives dimention as a parameter.
And divides it in half based on dimention , then multiplies the first half of the divided input by the second half to which sigmoid is applied.
Since the work we do is creating replicas of pytorch or tensorflow, be sure to refer to the definitions of pytorch and tensorflow. (https://pytorch.org/docs/stable/generated/torch.nn.GLU.html) 
The API shared externally in miopen.h is used directly by users. Do not change the definition except in some unavoidable cases.
Therefore, in order to implement GLU in the same way as pytorch, you must understand tensor view and slice. See https://github.com/ROCm/MIOpen/pull/2883. Please note that this PR is currently under review, so some details may change.

Comment (cognaiger9) [reply include/miopen/miopen.h:6604]: I've updated the interface to accept only one input, and it operates as follows:
```
    const size_t gid = blockIdx.x * blockDim.x + threadIdx.x;
    if(gid >= N)
        return;

    size_t innerIdx       = gid % inner_size;
    size_t splittedDimIdx = ((gid - innerIdx) / inner_size) % splitedDim_size;
    size_t outerIdx =
        (gid - innerIdx - splittedDimIdx * inner_size) / (inner_size * splitedDim_size);
    size_t inputIdx1 =
        outerIdx * splitDim_size * inner_size + splittedDimIdx * inner_size + innerIdx;
    size_t inputIdx2 = outerIdx * splitDim_size * inner_size +
                       (splittedDimIdx + splitedDim_size) * inner_size + innerIdx;

    FLOAT_ACCUM val1 = CVT_FLOAT2ACCUM(input[inputIdx1]);
    FLOAT_ACCUM val2 = sigmoid(CVT_FLOAT2ACCUM(input[inputIdx2]));
    FLOAT_ACCUM val  = val1 * val2;
    output[gid]      = CVT_ACCUM2FLOAT(val);
```
For each element in the output, I locate the corresponding element in the two halves of the input, multiply the first element by sigmoid of the second. Could you please review this implementation to ensure it meets the requirements?

File: src/kernels/MIOpenGLU.cpp (hunk 9 above)
Comment (kyeonghwanryu) [reply src/kernels/MIOpenGLU.cpp:45]: Yes it's my mistake. Using FLOAT_ACCUM is right. Sorry for that.

File: include/miopen/miopen.h (hunk 11 above)
Comment (seungmanhan) [reply include/miopen/miopen.h:6604]: If the dim being split is not the first dim, the tensor will be uncontiguous. If the tensor is non-contiguous your code will run into problems. Because it is not stored in memory in the correct order. 
Let's not work on non-contiguous, tensor view, and slice altogether now.
In the case of the current work, we are only working on contiguous, so work on the code assuming that only the first dim is spliced.

Comment (seungmanhan) [reply include/miopen/miopen.h:6604]: If you don't understand, look up tensor contiguous, non-contiguous, slicing.

Comment (cognaiger9) [reply include/miopen/miopen.h:6604]: I have changed my implementation as your comment (assuming only the first dim is spliced). Could you please check it?
```
template <typename TI, typename TO>
__device__ void
GLUFwdCOntiguousKernel(const TI* __restrict__ input, TO* __restrict__ output, long N)
{
    const TI* inputFirstHalf  = input;
    const TI* inputSecondHalf = input + N;
    const size_t gid          = blockIdx.x * blockDim.x + threadIdx.x;
    if(gid >= N)
        return;

    FLOAT_ACCUM val1 = CVT_FLOAT2ACCUM(inputFirstHalf[gid]);
    FLOAT_ACCUM val2 = sigmoid(CVT_FLOAT2ACCUM(inputSecondHalf[gid]));
    FLOAT_ACCUM val  = val1 * val2;
    output[gid]      = CVT_ACCUM2FLOAT(val);
}
```

File: test/cpu_glu.hpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+#include "tensor_holder.hpp"
+
+template <typename T>
+T sigmoid(T x)
+{
+    return static_cast<T>(1.0f / (1.0f + exp(-x)));
+}
+
+template <class T>
+void cpu_glu_forward(tensor<T> input, tensor<T>& ref_output, int dim)
```
Comment (seungmanhan) [inline test/cpu_glu.hpp]: Please also change this function appropriately.

File: include/miopen/miopen.h (hunk 11 above)
Comment (seungmanhan) [reply include/miopen/miopen.h:6604]: [resolved]

File: test/cpu_glu.hpp (hunk 12 above)
Comment (cognaiger9) [reply test/cpu_glu.hpp]: I have changed this function impl

**Comment by kyeonghwanryu** [comment]
Good work. You can change your ticket to done.
IsApplicable condition needs to be narrowed down to guarantee the better performance. Think about it while you do the next work.
